File: app/src/main/java/me/devsaki/hentoid/json/JsonCustomGroup.java
package me.devsaki.hentoid.json;

import androidx.annotation.NonNull;

import me.devsaki.hentoid.database.domains.Group;
import me.devsaki.hentoid.enums.Grouping;

class JsonCustomGroup {

    private String name;
    private Integer order;
    private Integer subtype;
    private Boolean favourite;
    private Integer rating;
    private Boolean hasCustomBookOrder;
    private String searchUri;

    private JsonCustomGroup() {
    }

    static JsonCustomGroup fromEntity(Group g) {
        JsonCustomGroup result = new JsonCustomGroup();
        result.name = g.name;
        result.order = g.order;
        result.subtype = g.subtype;
        result.favourite = g.isFavourite();
        result.rating = g.getRating();
        result.hasCustomBookOrder = g.hasCustomBookOrder;
        result.searchUri = g.searchUri;
        return result;
    }

    Group toEntity(@NonNull final Grouping grouping) {
        return new Group(grouping, name, order)
                .setSubtype((null == subtype) ? 0 : subtype)
                .setFavourite(null != favourite && favourite)
                .setRating(null == rating ? 0 : rating)
                .setHasCustomBookOrder(null != hasCustomBookOrder && hasCustomBookOrder)
                .setSearchUri(null == searchUri ? "" : searchUri);
    }
}

File: app/src/main/java/me/devsaki/hentoid/json/JsonCustomGroup.kt
package me.devsaki.hentoid.json

import com.squareup.moshi.JsonClass
import me.devsaki.hentoid.database.domains.Group
import me.devsaki.hentoid.enums.Grouping

@JsonClass(generateAdapter = true)
data class JsonCustomGroup(
    val name: String,
    val order: Int,
    val subtype: Int?,
    val favourite: Boolean?,
    val rating: Int?,
    val hasCustomBookOrder: Boolean?,
    val searchUri: String?
) {
    constructor(g: Group) : this(
        g.name,
        g.order,
        g.subtype,
        g.favourite,
        g.rating,
        g.hasCustomBookOrder,
        g.searchUri
    )

    fun toEntity(grouping: Grouping): Group {
        return Group(
            grouping = grouping,
            name = name,
            order = order,
            subtype = subtype ?: 0,
            favourite = favourite ?: false,
            rating = rating ?: 0,
            hasCustomBookOrder = hasCustomBookOrder ?: false,
            searchUri = searchUri ?: ""
        )
    }
}